src/router/validation/subscriptionValidator.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+import Ajv from 'ajv';
+import { isEmpty, groupBy } from 'lodash';
+
+import { InvalidResourceError, Validator, Search } from 'fhir-works-on-aws-interface';
+import subscriptionSchema from './subscriptionSchema.json';
+
+export interface SubscriptionEndpoint {
+    endpoint: string | RegExp;
+    headers?: string[];
+    tenantId?: string;
+}
+
+export type GetAllowEndpoints = () => Promise<SubscriptionEndpoint[]>;
+
+const SUBSCRIPTION_RESOURCE_TYPE = 'Subscription';
+
+const BUNDLE_RESOURCE_TYPE = 'Bundle';
+
+const SINGLE_TENANT_ALLOW_LIST_KEY = 'SINGLE_TENANT_ALLOW_LIST_KEY';
+
+export default class SubscriptionValidator implements Validator {
+    private ajv: Ajv.Ajv;
+
+    private validateJSON: Ajv.ValidateFunction;
+
+    private search: Search;
+
+    private allowListMap: { [key: string]: (string | RegExp)[] } = {};
+
+    private allowListLoaded = false;
```

Comment:
I think it would be simpler to just have `SubscriptionValidator` receive `SubscriptionEndpoint[]` as param instead of having to call a fn and caching the result.

That way the client (most likely the `config.ts` file in fwoa-deployment) can fetch the `SubscriptionEndpoint` array and construct the Validator. In the default setup the allow list can be fetched just once at server startup(when building FhirConfig).

Maybe I'm missing something, is there a reason to have the fn call in this class?

Reply:
Yes, it would be simpler to just pass in the endpoints. I didn't do it as there's no `await` used in `config.ts`, and I thought it was done on purpose to avoid top-level await problems. Maybe I'm overthinking this?

Reply:
`await` won't be a problem. A few months ago we made `getFhirConfig` async so that customers can build their config asynchronously. The `index.ts` already does `await getFhirConfig()` (even if by default all config is static)

https://github.com/awslabs/fhir-works-on-aws-deployment/pull/464